### DISCUSSION TITLE
Chakra 3: Snippets? what is this?

### DISCUSSION THREAD

**Discussion by robot-ranger** [chakra-ui/chakra-ui#9295] [Q&A] [answered]
Someone please help. I was just getting comfortable with chakra, starting from https://github.com/fastapi/full-stack-fastapi-template.

Ive been trying to migrate to v3, but theres a ton of imports missing. That seem like basic components such as InputGroup, Drawer, useColorMode, etc.

I guess i need to import some of these as snippets? what is this? Someone please help me understand the new architecture of chakra because im fully confused and the docs dont really help.

**Comment by veloware** [comment, marked as answer]
`@chakra-ui/react` v3 contains primitive components, such as Box, HStack etc, these are all simple, atomic components.  Snippets are pre-built components that have a bit more functionality, that compose of other components.  You'll need to check the docs for your components to see if they need imported as a primitive from `@chakra-ui/react` e.g. `import { Box } from '@chakra-ui/react` or if they need added as a snippet.  Adding a snippet will add the snippet code to your workspace, when using the snippet, you will import it from your own workspace - `import { Button } from '@/components/button'`

There is confusion around similarly named primitives and snippets though, which there are other discussions on, for example 'button'.  There is a button from chakra as a primitive, and also as a snippet.  In this case it's the snippet you want to use, as it will handle loading states etc.

If you add all or some of the snippets to your workspace, have a look inside them and you'll hopefully see how the primitives are brought in from Chakra and enhanced with more functionality.  Here's the button snippet as of today, as an example - 

```
import type { ButtonProps as ChakraButtonProps } from "@chakra-ui/react"
import {
  AbsoluteCenter,
  Button as ChakraButton,
  Span,
  Spinner,
} from "@chakra-ui/react"
import { forwardRef } from "react"

interface ButtonLoadingProps {
  loading?: boolean
  loadingText?: React.ReactNode
}

export interface ButtonProps extends ChakraButtonProps, ButtonLoadingProps {}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(props, ref) {
    const { loading, disabled, loadingText, children, ...rest } = props
    return (
      <ChakraButton disabled={loading || disabled} ref={ref} {...rest}>
        {loading && !loadingText ? (
          <>
            <AbsoluteCenter display="inline-flex">
              <Spinner size="inherit" color="inherit" />
            </AbsoluteCenter>
            <Span opacity={0}>{children}</Span>
          </>
        ) : loading && loadingText ? (
          <>
            <Spinner size="inherit" color="inherit" />
            {loadingText}
          </>
        ) : (
          children
        )}
      </ChakraButton>
    )
  },
)

```

The benefit is, it gives you flexibility to adjust the snippets to suit the needs of your application. Whereas before, it would be hard to make a closed component of a 'Button' to suit everyone's needs.

I've already started a discussion on how to possibly improve the docs to distinguish the difference, as it is quite confusing - https://github.com/chakra-ui/chakra-ui/discussions/9143

**Comment by robot-ranger** [comment]
Thank you that is very helpful! I appreciate the explaination.

I was also fighting https://github.com/npm/cli/issues/7814, and I was incorrectly conflating npm hanging up with chakra.
...so now that I have updated npm to 10.9.1... the npm cli actually works and therefore the chakra cli.

Its making more sense now, and ive got my buttons moved over to snippets.

I was certainly having trouble telling the difference between snippet-provided components and 'chakra-native' components in the docs. I have now started to pay attention to the import paths shown, looking for `'@/components/ui/...'` to indicate that as a snippet-provided component, as opposed to `'@chakra-ui/react` for chakra-native components

Frankly, I dont like the name 'snippets'. Of course, 'what is a name?', i can get over it. But at the same time - I do associate snippets with things like VSCode plugins akin to 'Autocomplete Snippets'. These new Chakra v3 snippets are kinda the same, but they feel different when part of a UI package like chakra for some reason.

RE: Documentation: IMO, due to the significant differences in workflow to get a snippet on the screen vs a chakra-native component on the screen (the need for chakra cli), they should not be lumped together in the docs. I think it would be better to document all of the chakra-native components in one place and separate tab/section for snippets:

For Example:
![image](https://github.com/user-attachments/assets/107c61bc-577c-47db-b7cf-43bfc2cc356c)

Take that as a grain of salt, just some raw thoughts from me, some dude on the internet.

**Comment by robot-ranger** [comment]
Some snippets use elements in PascalCase, and some use dot notation 

Ex:
`<MenuRoot>` vs `<Card.Root>`

...Why?

**Reply by veloware** [reply]
Not quite, in your case above, `Card` is imported from`@chakra-ui/react`, where `Card` is the namespace that contains the various parts of a compound component, e.g. a `Card` can have `Root`, `Body` and so on, but each of those are probably just `div` elements with styles attached. 

However if you look at the Menu **snippet**, you can see there are more complex components such as MenuRadioItem, where MenuRadioItem will be composed of other components/primitives (using the dot notation).

`MenuRoot` though, to your example, if you look, just re-exports `Menu.Root`, but it is one of many parts of a overall Menu snippet.

In essence, the dot notation is just a way to namespace components, but it looks to be used more with the component parts from Chakra, whereas the snippet sub-components seem to keep the full name.

Namespacing also simplifies the import, as you only need to import the namespace, and not each individual sub component

Thanks my understanding anyway...

**Comment by ftzi** [comment]
Yes snippets are a terrible thing for me. Deal breaker. Wanted to move out of Mantine but it seems I will keep with it.

If I want to have my own components I would use shadcn, but I don't want to. I just want to import them.

Why not include all of them by default and allow via CLI to copy their implementation to the project, to allow customization?

**Reply by robot-ranger** [reply]
I'm relatively new to all of this, I don't have years of experience in various UI design systems but my best understanding is that a UI library =\= component library. It's impossible to satisfy everyone.

If you are ok with 'locked in' preconfigured components and you just want to import and go and lean into the aesthetic provided by a given component library, then I guess chakra is not for you (anymore?).

For me, I've just imported and not customized I just 'import and go'.

Now that I know more about what and how to use snippets, I'm ok with it and I appreciate being able to make local customizations.

I will admit that even opening a terminal to run the cli commands is a hiccup in the workflow. It feels out of place. We use devcontainers that do all the setup and launching of the dev server. We hardly typically touch the terminal so the cli is new muscle memory for us

**Reply by veloware** [reply]
To be fair, if you import all the snippets with the single command and don't alter them, you have kind of the same experience as before, as in, pre-made components that you can import and use (you're just importing them from a different place)

The issues might start if you make large updates to those local snippets, and then try and keep them up to date with any changes from the Chakra side, as there would be a deviation between the two, it'll be up to you to manage those deviations.  This wasn't possible in v2 though, as you couldn't alter the components from Chakra.

**Reply by robot-ranger** [reply]
eh... sort of... 

I suppose the high-level workflow is the same: "import" the bit, then use it. Only now we have to do BOTH `cli snippet add` AND `import { .. } from "...../ui/......"

its fine i guess, but it is an extra step.

Speaking of keeping up to date - how do we do that? do we just run the `snippet add ....` again?
